// Copyright 2025 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package wfa.measurement.reporting.v2alpha;

import "google/api/field_behavior.proto";

option java_package = "org.wfanet.measurement.reporting.v2alpha";
option java_multiple_files = true;
option java_outer_classname = "EventTemplateFieldProto";
option go_package = "github.com/world-federation-of-advertisers/cross-media-measurement/reporting/apiv2alpha/reportingpb";

// A representation of an event template field with its value
message EventTemplateField {
  // A field path with respect to the local market Event message
  // with `.` as the traversal operator.
  //
  // For example, if the Event message had an event template `common` with
  // a field `age_group` then the field path would be `common.age_group`.
  //
  // Only population attributes are supported at the moment.
  string path = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.field_behavior) = IMMUTABLE
  ];
  // The value of the event template field
  message FieldValue {
    // Value type selector. Required.
    oneof selector {
      // The field value as a string. Only STRING field type is supported.
      string string_value = 1 [(google.api.field_behavior) = IMMUTABLE];
      // The field value as a stringified enum value. This is the symbol
      // associated with the enum value, not a stringification of its integer
      // value. For example, for MediaType.VIDEO we expect "VIDEO" not "1".
      // Only ENUM field type is supported.
      string enum_value = 2 [(google.api.field_behavior) = IMMUTABLE];
      // The field value as a Boolean. Only BOOL field type is supported.
      bool bool_value = 3 [(google.api.field_behavior) = IMMUTABLE];
      // The field value as a Float. Only FLOAT field type is supported.
      float float_value = 4 [(google.api.field_behavior) = IMMUTABLE];
    }
  }
  // The value of the field
  FieldValue value = 2 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.field_behavior) = IMMUTABLE
  ];
}
